src/main/java/de/tum/cit/aet/artemis/iris/web/IrisChatSessionResource.java

```diff
@@ -0,0 +1,252 @@
+                        .orElseThrow();
+                programmingExerciseChatSession.setEntityId(programmingExerciseChatSession.getExerciseId());
+                session = programmingExerciseChatSession;
+            }
+        }
+
+        return ResponseEntity.ok(session);
+    }
+
+    /**
+     * GET chat-history/{courseId}/sessions: Retrieve all Iris Sessions for the course
+     *
+     * @param courseId of the course
+     * @return the {@link ResponseEntity} with status {@code 200 (Ok)} and with body a list of the iris sessions for the course or {@code 404 (Not Found)} if no session exists
+     */
+    @GetMapping("{courseId}/sessions")
+    @EnforceAtLeastStudentInCourse
+    public ResponseEntity<List<IrisChatSession>> getAllSessionsForCourse(@PathVariable Long courseId) {
+        var user = userRepository.getUserWithGroupsAndAuthorities();
+        if (user.hasAcceptedExternalLLMUsage()) {
+            var allChatSessions = Stream.of(getAllSessionsForCourseChat(courseId), getAllSessionsForLectureChat(courseId), getAllSessionsForProgrammingExerciseChat(courseId),
+                    getAllSessionsForTextExerciseChat(courseId)).flatMap(List::stream).toList();
+            return ResponseEntity.ok(allChatSessions);
+        }
+        else {
+            return ResponseEntity.ok(List.of());
+        }
+    }
+
+    private List<IrisChatSession> getAllSessionsForCourseChat(Long courseId) {
+        var course = courseRepository.findById(courseId);
+        if (course.isPresent()) {
+            if (irisSettingsService.isEnabledFor(IrisSubSettingsType.COURSE_CHAT, course.get())) {
+                var user = userRepository.getUserWithGroupsAndAuthorities();
+
+                var sessions = irisCourseChatSessionRepository.findLatestByCourseIdAndUserIdWithMessages(course.get().getId(), user.getId(), Pageable.unpaged());
+                sessions.forEach(s -> {
+                    irisSessionService.checkHasAccessToIrisSession(s, user);
+                    s.setEntityId(s.getCourseId());
+                });
+                return new ArrayList<>(sessions);
+            }
+        }
+        return Collections.emptyList();
+    }
+
+    private List<IrisChatSession> getAllSessionsForLectureChat(Long courseId) {
+        var course = courseRepository.findById(courseId);
+
+        if (course.isPresent()) {
+            Set<Lecture> lecturesForCourse = lectureRepositoryApi.findAllByCourseId(courseId);
+
+            if (irisSettingsService.isEnabledFor(IrisSubSettingsType.LECTURE_CHAT, course.get())) {
+                var user = userRepository.getUserWithGroupsAndAuthorities();
+                List<IrisLectureChatSession> sessions = lecturesForCourse.stream()
+                        .flatMap(l -> irisLectureChatSessionRepository.findLatestSessionsByLectureIdAndUserIdWithMessages(l.getId(), user.getId(), Pageable.unpaged()).stream())
```

Comment:
please avoid calling a repository method in a stream. This is similar to a for loop, so for every lecture, the latest sessions are retrieved